fabric-events-interaction-v0/src/main/java/net/fabricmc/fabric/mixin/event/interaction/MixinServerPlayerInteractionManager.java

```diff
@@ -81,4 +82,28 @@ public void interactItem(ServerPlayerEntity player, World world, ItemStack stack
 			return;
 		}
 	}
+
+	@Inject(at = @At(value = "INVOKE", target = "Lnet/minecraft/block/Block;onBreak(Lnet/minecraft/world/World;Lnet/minecraft/util/math/BlockPos;Lnet/minecraft/block/BlockState;Lnet/minecraft/entity/player/PlayerEntity;)V"), method = "tryBreakBlock", locals = LocalCapture.CAPTURE_FAILHARD, cancellable = true)
+	private void breakBlock(BlockPos pos, CallbackInfoReturnable<Boolean> cir, BlockState state, BlockEntity entity, Block block) {
+		ActionResult result = BlockBreakEvents.BEFORE.invoker().beforeBlockBreak(this.player, pos, state, entity, block);
+
+		if (result == ActionResult.FAIL) {
+			BlockPos cornerPos = pos.add(-1, -1, -1);
```

Comment:
Would a mutable block pos be possible to use here if it's beneficial for performance.

Reply:
I don't know if it's possible, as the packet might change as the loop iterates

Reply:
Unfortunately cannot, as the packet is lazily encoded.  This really is a garbage factory as-is, with all the block positions and packets.  Would consider using a custom packet that does the iteration client-side.

This *probably* won't matter in single player with mostly vanilla content, but in a multi-player environment with expanded-area tools or quarries that impersonate players it could become A Problem.

Reply:
Even in modded I doubt it, even at 1 block/tick that's like 27 Objects per tick which is pretty much nothing compared to the other stuff vanilla does :p

Reply:
1 block per tick? How about a Project E Destruction Catalyst or Botania TerraSteel Pick on max settings?

Reply:
1 block per tick? How about a Project E Destruction Catalyst or Botania TerraSteel Pick on max settings?

Reply:
I think the reason for these packets is to tell the client what the neighboring blocks actually are (since the client also calculates this).

Reply:
shouldn't the block and it's neighbor not change if it is failing?